src/slip.rs

```diff
@@ -0,0 +1,200 @@
+use secp256k1::{PublicKey};
+
+/// A UTXO slip containing record of owernship of funds on the network
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct Slip {
+    /// Contains concrete data which is not relative to state of the chain
+    body: SlipBody,
+    /// Flag determining if the `Slip` is part of the longest chain
+    lc: bool,
+    /// `SlipSpentStatus` indicating the state of the slip in the network
+    pub spent_status: SlipSpentStatus,
+}
+
+/// An object that holds concrete data not subjective to state of chain
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct SlipBody {
+    /// An `Sectp256K::PublicKey` determining who owns the `Slip`
+    add: PublicKey,
+    /// A enum brodcast type determining how to be processed by consensus
+    typ: SlipBroadcastType,
+    /// Amount of Saito
+    amt: u64,
+    /// `Block` id
+    bid: u64,
+    /// `Transaction` id
+    tid: u64,
+    /// `Slip` id
+    sid: u64,
+    /// The `Block` hash the slip originated from
+    bsh: [u8; 32],
+}
+
+/// An enumerated set of `Slip` types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum SlipBroadcastType {
+    Normal,
+    GoldenTicket,
+    Fee,
+    Rebroadcast,
+    VIP,
+    GoldenChunk,
+}
+
+/// Current status of the `Slip` state with regards to it's usage in the network
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum SlipSpentStatus {
+    Unspent,
+    Spent,
+    Pending,
+}
+
+impl Slip {
+    /// Create new `Slip` with default type `SlipBroadcastType::Normal`
+    ///
+    /// * `add` - `Publickey` address to assign ownership
+    /// * `typ` - `SlipBroadcastType` of `Slip`
+    /// * `amt` - `u64` amount of Saito coantined in the `Slip`
+    pub fn new(add: PublicKey, typ: SlipBroadcastType, amt: u64) -> Slip {
+        return Slip {
+            body: SlipBody {
+                add,
+                typ,
+                amt,
+                bid: 0,
+                tid: 0,
+                sid: 0,
+                bsh: [0; 32],
+            },
+            lc: false,
+            spent_status: SlipSpentStatus::Unspent,
+        }
+    }
+
+    /// Returns address in `Slip`
+    pub fn get_address(&self) -> PublicKey {
+        return self.body.add;
+    }
+
+    /// Returns`Slip` type from the enumerated set of `SlipBroadcastType`
+    pub fn get_type(&self) -> SlipBroadcastType {
+        return self.body.typ.clone();
```

Comment:
Does this need to be clone()ed?

Reply:
I believe so, let me see if I can implement the `Copy` trait on it instead

Reply:
Just so it's clear, this is not socratic method, i actually have no idea if it should be clone()ed or not.